components/ws-manager/pkg/manager/monitor.go

```diff
@@ -927,7 +928,122 @@ func (m *Monitor) finalizeWorkspaceContent(ctx context.Context, wso *workspaceOb
 		}
 
-		if doSnapshot {
+		if pvcFeatureEnabled {
+			pvcName := wso.Pod.Name
+			if !createdVolumeSnapshot {
+				// create snapshot object out of PVC
+				volumeSnapshot := &volumesnapshotv1.VolumeSnapshot{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      pvcVolumeSnapshotName,
+						Namespace: m.manager.Config.Namespace,
+						Labels: map[string]string{
+							"workspaceID": workspaceID,
+						},
+					},
+					Spec: volumesnapshotv1.VolumeSnapshotSpec{
+						Source: volumesnapshotv1.VolumeSnapshotSource{
+							PersistentVolumeClaimName: &pvcName,
+						},
+						VolumeSnapshotClassName: &pvcVolumeSnapshotClassName,
+					},
+				}
+
+				err = m.manager.Clientset.Create(ctx, volumeSnapshot)
+				if err != nil {
+					err = xerrors.Errorf("cannot create volumesnapshot: %v", err)
+					return true, nil, err
+				}
+				createdVolumeSnapshot = true
+			}
+			if createdVolumeSnapshot {
+				backoff := wait.Backoff{
+					Steps:    30,
+					Duration: 100 * time.Millisecond,
+					Factor:   1.5,
+					Jitter:   0.1,
+					Cap:      10 * time.Minute,
+				}
+				err = wait.ExponentialBackoff(backoff, func() (bool, error) {
+					var volumeSnapshot volumesnapshotv1.VolumeSnapshot
+					err := m.manager.Clientset.Get(ctx, types.NamespacedName{Namespace: m.manager.Config.Namespace, Name: pvcVolumeSnapshotName}, &volumeSnapshot)
+					if err != nil {
+						if k8serr.IsNotFound(err) {
+							// volumesnapshot doesn't exist yet, retry again
+							return false, nil
+						}
+						log.WithError(err).WithField("VolumeSnapshot.Name", pvcVolumeSnapshotName).Error("was unable to get volume snapshot")
+						return false, err
+					}
+					if volumeSnapshot.Status != nil {
+						if volumeSnapshot.Status.ReadyToUse != nil && *(volumeSnapshot.Status.ReadyToUse) {
+							pvcVolumeSnapshotContentName = *volumeSnapshot.Status.BoundVolumeSnapshotContentName
+							return true, nil
+						}
+						if volumeSnapshot.Status.Error != nil {
+							if volumeSnapshot.Status.Error.Message != nil {
+								err = xerrors.Errorf("error during volume snapshot creation: %s", *volumeSnapshot.Status.Error.Message)
+								log.WithError(err).WithField("VolumeSnapshot.Name", pvcVolumeSnapshotName).Error("unable to create volume snapshot")
+								return false, err
+							}
+							log.WithField("VolumeSnapshot.Name", pvcVolumeSnapshotName).Error("unknown error during volume snapshot creation")
+							return false, xerrors.Errorf("unknown error during volume snapshot creation")
+						}
+					}
+					return false, nil
+				})
+				if err != nil {
+					log.WithError(err).WithField("VolumeSnapshot.Name", pvcVolumeSnapshotName).Errorf("failed while waiting for volume snapshot to get ready")
+					return true, nil, err
+				}
+				readyVolumeSnapshot = true
+			}
+			if readyVolumeSnapshot && !markVolumeSnapshotAnnotation {
+				var volumeSnapshotContent volumesnapshotv1.VolumeSnapshotContent
```

Comment:
We could add the VolumeSnapshotContentName in the log at this entrypoint, so the following logs, we don't have to add `.WithField("VolumeSnapshotContent.Name", pvcVolumeSnapshotContentName)`
```go
log = log.WithField("VolumeSnapshotContent.Name", pvcVolumeSnapshotContentName)
```